crates/shared/src/oneinch_api.rs

```diff
@@ -384,6 +386,55 @@ where
     serde_json::from_str(&response?).context("1inch result parsing failed")
 }
 
+#[derive(Debug, Clone)]
+pub struct ProtocolCache(Arc<Mutex<TimedSizedCache<Vec<String>, Vec<String>>>>);
+
+impl ProtocolCache {
+    pub fn new(cache_size: usize, cache_validity_in_seconds: u64) -> Self {
+        Self(Arc::new(Mutex::new(
+            TimedSizedCache::with_size_and_lifespan_and_refresh(
+                cache_size,
+                cache_validity_in_seconds,
+                false,
+            ),
+        )))
+    }
+
+    pub async fn get_allowed_protocols(
+        &self,
+        #[allow(clippy::ptr_arg)] disabled_protocols: &Vec<String>,
+        api: &dyn OneInchClient,
+    ) -> Result<Option<Vec<String>>> {
+        if disabled_protocols.is_empty() {
+            return Ok(None);
+        }
+
+        if let Some(allowed_protocols) = self.0.lock().unwrap().cache_get(disabled_protocols) {
```

Comment:
Interesting... This doesn't use `Borrow::Target` so you have to use a `&Vec<String>` instead of the more Rust-y `&[String]`.